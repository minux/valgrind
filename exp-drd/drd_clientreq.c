/*
  This file is part of drd, a data race detector.

  Copyright (C) 2006-2008 Bart Van Assche
  bart.vanassche@gmail.com

  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU General Public License as
  published by the Free Software Foundation; either version 2 of the
  License, or (at your option) any later version.

  This program is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
  02111-1307, USA.

  The GNU General Public License is contained in the file COPYING.
*/


#include "drd_clientreq.h"
#include "drd_cond.h"
#include "drd_mutex.h"
#include "drd_semaphore.h"
#include "drd_suppression.h"      // drd_start_suppression()
#include "drd_thread.h"
#include "drd_track.h"
#include "priv_drd_clientreq.h"
#include "pub_tool_basics.h"      // Bool
#include "pub_tool_libcassert.h"
#include "pub_tool_libcassert.h"  // tl_assert()
#include "pub_tool_libcprint.h"   // VG_(message)()
#include "pub_tool_machine.h"     // VG_(get_SP)()
#include "pub_tool_threadstate.h"
#include "pub_tool_tooliface.h"   // VG_(needs_...)()


static void drd_spin_init_or_unlock(const Addr spinlock, const SizeT size)
{
   struct mutex_info* mutex_p = mutex_get(spinlock);
   if (mutex_p)
   {
      mutex_unlock(spinlock, mutex_type_spinlock);
   }
   else
   {
      mutex_init(spinlock, size, mutex_type_spinlock);
   }
}

static void drd_pre_cond_wait(const Addr cond, const SizeT cond_size,
                              const Addr mutex, const MutexT mutex_type)
{
   mutex_unlock(mutex, mutex_type);
   cond_pre_wait(cond, cond_size, mutex);
}

static void drd_post_cond_wait(const Addr cond, const Addr mutex,
                               const SizeT size, const MutexT mutex_type)
{
   cond_post_wait(cond);
   mutex_post_lock(mutex, size, mutex_type);
}

static void drd_pre_cond_signal(const Addr cond)
{
   cond_pre_signal(cond);
}

static void drd_pre_cond_broadcast(const Addr cond)
{
   cond_pre_broadcast(cond);
}

static Bool drd_handle_client_request(ThreadId tid, UWord* arg, UWord* ret)
{
   UWord result = 0;

   switch (arg[0])
   {
   case VG_USERREQ__GET_THREAD_SELF:
      result = tid;
      break;

   case VG_USERREQ__SET_THREAD_NAME:
      thread_set_name_fmt(VgThreadIdToDrdThreadId(VG_(get_running_tid)()),
                          (char*)arg[1], arg[2]);
      break;

   case VG_USERREQ__DRD_START_SUPPRESSION:
      drd_start_suppression(arg[1], arg[1] + arg[2], "client");
      break;

   case VG_USERREQ__DRD_FINISH_SUPPRESSION:
      drd_finish_suppression(arg[1], arg[1] + arg[2]);
      break;

   case VG_USERREQ__DRD_SUPPRESS_CURRENT_STACK:
      thread_set_stack_startup(thread_get_running_tid(),
                               VG_(get_SP)(VG_(get_running_tid)()));
      break;

   case VG_USERREQ__DRD_START_NEW_SEGMENT:
      thread_new_segment(PtThreadIdToDrdThreadId(arg[1]));
      break;

   case VG_USERREQ__DRD_START_RECORDING:
      thread_start_recording(PtThreadIdToDrdThreadId(arg[1]));
      break;

   case VG_USERREQ__DRD_STOP_RECORDING:
      thread_stop_recording(PtThreadIdToDrdThreadId(arg[1]));
      break;

   case VG_USERREQ__SET_PTHREADID:
      thread_set_pthreadid(thread_get_running_tid(), arg[1]);
      break;

   case VG_USERREQ__SET_JOINABLE:
      thread_set_joinable(PtThreadIdToDrdThreadId(arg[1]), (Bool)arg[2]);
      break;

   case VG_USERREQ__POST_THREAD_JOIN:
      tl_assert(arg[1]);
      drd_post_thread_join(thread_get_running_tid(),
                           PtThreadIdToDrdThreadId(arg[1]));
      break;

   case VG_USERREQ__PRE_MUTEX_INIT:
      drd_pre_mutex_init(arg[1], arg[2], arg[3]);
      break;

   case VG_USERREQ__POST_MUTEX_DESTROY:
      drd_post_mutex_destroy(arg[1], arg[2]);
      break;

   case VG_USERREQ__PRE_PTHREAD_MUTEX_LOCK:
      drd_pre_mutex_lock(thread_get_running_tid(), arg[1], arg[2], arg[3]);
      break;

   case VG_USERREQ__POST_PTHREAD_MUTEX_LOCK:
      drd_post_mutex_lock(thread_get_running_tid(), arg[1], arg[2], arg[3]);
      break;

   case VG_USERREQ__PRE_PTHREAD_MUTEX_UNLOCK:
      drd_pre_mutex_unlock(thread_get_running_tid(), arg[1], arg[3]);
      break;

   case VG_USERREQ__SPIN_INIT_OR_UNLOCK:
      drd_spin_init_or_unlock(arg[1], arg[2]);
      break;

   case VG_USERREQ__POST_PTHREAD_COND_INIT:
      drd_post_cond_init(arg[1], arg[2]);
      break;

   case VG_USERREQ__PRE_PTHREAD_COND_DESTROY:
      drd_pre_cond_destroy(arg[1]);
      break;

   case VG_USERREQ__PRE_PTHREAD_COND_WAIT:
      drd_pre_cond_wait(arg[1]/*cond*/,  arg[2]/*cond_size*/,
                        arg[3]/*mutex*/, arg[4]/*mutex_type*/);
      break;

   case VG_USERREQ__POST_PTHREAD_COND_WAIT:
      drd_post_cond_wait(arg[1]/*cond*/, arg[2]/*mutex*/,
                         arg[3]/*mutex_size*/, arg[4]/*mutex_type*/);
      break;

   case VG_USERREQ__PRE_PTHREAD_COND_SIGNAL:
      drd_pre_cond_signal(arg[1]);
      break;

   case VG_USERREQ__PRE_PTHREAD_COND_BROADCAST:
      drd_pre_cond_broadcast(arg[1]);
      break;

   case VG_USERREQ__SEM_INIT:
      drd_semaphore_init(arg[1], arg[2], arg[3], arg[4]);
      break;

   case VG_USERREQ__SEM_DESTROY:
      drd_semaphore_destroy(arg[1]);
      break;

   case VG_USERREQ__POST_SEM_WAIT:
      drd_semaphore_post_wait(thread_get_running_tid(), arg[1], arg[2]);
      break;

   case VG_USERREQ__PRE_SEM_POST:
      drd_semaphore_pre_post(thread_get_running_tid(), arg[1], arg[2]);
      break;

   case VG_USERREQ__POST_SEM_POST:
      drd_semaphore_post_post(thread_get_running_tid(), arg[1], arg[2], arg[3]);
      break;

   case VG_USERREQ__BARRIER_INIT:
      drd_barrier_init(arg[1], arg[2], arg[3]);
      break;

   case VG_USERREQ__BARRIER_DESTROY:
      drd_barrier_destroy(arg[1]);
      break;

   case VG_USERREQ__PRE_BARRIER_WAIT:
      drd_barrier_pre_wait(thread_get_running_tid(), arg[1]);
      break;

   case VG_USERREQ__POST_BARRIER_WAIT:
      drd_barrier_post_wait(thread_get_running_tid(), arg[1], arg[2]);
      break;

   default:
      VG_(message)(Vg_DebugMsg, "Unrecognized client request 0x%lx 0x%lx",
                   arg[0], arg[1]);
      tl_assert(0);
      return False;
   }

   *ret = result;
   return True;
}

void drd_clientreq_init(void)
{
   VG_(needs_client_requests)(drd_handle_client_request);
}

/*
 * Local variables:
 * c-basic-offset: 3
 * End:
 */
