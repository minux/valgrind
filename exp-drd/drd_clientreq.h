#ifndef __DRD_CLIENTREQ_H
#define __DRD_CLIENTREQ_H


#include "valgrind.h" // VG_USERREQ_TOOL_BASE()


#define DRD_IGNORE_VAR(x) { int res; VALGRIND_DO_CLIENT_REQUEST(res, 0, VG_USERREQ__DRD_START_SUPPRESSION, &(x), sizeof(x), 0, 0, 0); }
#define DRD_TRACE_VAR(x)  { int res; VALGRIND_DO_CLIENT_REQUEST(res, 0, VG_USERREQ__DRD_START_TRACE_ADDR, &(x), sizeof(x), 0, 0, 0); }


enum {
  /* Ask the core the thread ID assigned by Valgrind. */
  VG_USERREQ__GET_THREAD_SELF = VG_USERREQ_TOOL_BASE('D', 'R'),
  /* args: none. */

  /* To tell the drd tool to suppress data race detection on the specified */
  /* address range. */
  VG_USERREQ__DRD_START_SUPPRESSION,
  /* args: start address, size in bytes */
  /* To tell the drd tool no longer to suppress data race detection on the */
  /* specified address range. */
  VG_USERREQ__DRD_FINISH_SUPPRESSION,
  /* args: start address, size in bytes */
  /* Ask drd to suppress data race reports on all currently allocated stack */
  /* data of the current thread.                                            */
  VG_USERREQ__DRD_SUPPRESS_CURRENT_STACK,
  /* args: none */
  /* To ask the drd tool to start a new segment in the specified thread. */
  VG_USERREQ__DRD_START_NEW_SEGMENT,
  /* args: POSIX thread ID. */
  /* To ask the drd tool to trace all accesses to the specified range. */
  VG_USERREQ__DRD_START_TRACE_ADDR,
  /* args: Addr, SizeT. */
  /* To ask the drd tool to stop tracing accesses to the specified range. */
  VG_USERREQ__DRD_STOP_TRACE_ADDR,
  /* args: Addr, SizeT. */
  /* Let the drd tool stop recording memory accesses in the calling thread. */
  VG_USERREQ__DRD_STOP_RECORDING,
  /* args: none. */
  /* Let the drd tool start recording memory accesses in the calling thread. */
  VG_USERREQ__DRD_START_RECORDING,
  /* args: none. */

  /* Tell the core the pthread_t of the running thread */
  VG_USERREQ__SET_PTHREADID,
  /* args: pthread_t. */
  /* Ask the core that a the thread's state transition from */
  /* VgTs_Zombie to VgTs_Empty is delayed until */
  /* VG_USERREQ__POST_THREAD_JOIN is performed. */
  VG_USERREQ__SET_JOINABLE,
  /* args: pthread_t, Bool */

  /* To notify drd that a thread finished because */
  /* pthread_thread_join() was called on it. */
  VG_USERREQ__POST_THREAD_JOIN,
  /* args: pthread_t (joinee) */

  /* to notify the drd tool of a pthread_mutex_init call. */
  VG_USERREQ__PRE_MUTEX_INIT,
  /* args: Addr, MutexT */
  /* to notify the drd tool of a pthread_mutex_init call. */
  VG_USERREQ__POST_MUTEX_INIT,
  /* args: Addr */
  /* to notify the drd tool of a pthread_mutex_destroy call. */
  VG_USERREQ__PRE_MUTEX_DESTROY,
  /* args: Addr */
  /* to notify the drd tool of a pthread_mutex_destroy call. */
  VG_USERREQ__POST_MUTEX_DESTROY,
  /* args: Addr, MutexT */
  /* to notify the drd tool of pthread_mutex_lock calls */
  VG_USERREQ__PRE_MUTEX_LOCK,
  /* args: Addr, MutexT, Bool */
  /* to notify the drd tool of pthread_mutex_lock calls */
  VG_USERREQ__POST_MUTEX_LOCK,
  /* args: Addr, Bool */
  /* to notify the drd tool of pthread_mutex_unlock calls */
  VG_USERREQ__PRE_MUTEX_UNLOCK,
  /* args: Addr */
  /* to notify the drd tool of pthread_mutex_unlock calls */
  VG_USERREQ__POST_MUTEX_UNLOCK,
  /* args: Addr */
  VG_USERREQ__SPIN_INIT_OR_UNLOCK,
  /* args: Addr spinlock */


  /* to notify the drd tool of a pthread_cond_init call. */
  VG_USERREQ__PRE_COND_INIT,
  /* args: Addr */
  /* to notify the drd tool of a pthread_cond_destroy call. */
  VG_USERREQ__POST_COND_DESTROY,
  /* args: Addr cond */
  VG_USERREQ__PRE_COND_WAIT,
  /* args: Addr cond, Addr mutex, MutexT mt */
  VG_USERREQ__POST_COND_WAIT,
  /* args: Addr cond, Addr mutex, Bool took_lock*/
  VG_USERREQ__PRE_COND_SIGNAL,
  /* args: Addr cond */
  VG_USERREQ__PRE_COND_BROADCAST,
  /* args: Addr cond */

  /* To notify the drd tool of a sem_init call. */
  VG_USERREQ__PRE_SEM_INIT,
  /* args: Addr sem, Word pshared, Word value */
  /* To notify the drd tool of a sem_init call. */
  VG_USERREQ__POST_SEM_INIT,
  /* args: Addr sem */
  /* To notify the drd tool of a sem_destroy call. */
  VG_USERREQ__PRE_SEM_DESTROY,
  /* args: Addr sem */
  /* To notify the drd tool of a sem_destroy call. */
  VG_USERREQ__POST_SEM_DESTROY,
  /* args: Addr sem */
  /* To notify the drd tool of a sem_wait call. */
  VG_USERREQ__PRE_SEM_WAIT,
  /* args: Addr sem */
  /* To notify the drd tool of a sem_wait call. */
  VG_USERREQ__POST_SEM_WAIT,
  /* args: Addr sem, Bool waited */
  /* To notify the drd tool before a sem_post call. */
  VG_USERREQ__PRE_SEM_POST,
  /* args: Addr sem */
  /* To notify the drd tool after a sem_post call. */
  VG_USERREQ__POST_SEM_POST,
  /* args: Addr sem, Bool waited */

  /* To notify the drd tool of a pthread_barrier_init call. */
  VG_USERREQ__PRE_BARRIER_INIT,
  /* args: Addr barrier, BarrierT type, Word count, Bool reinit */
  /* To notify the drd tool of a pthread_barrier_init call. */
  VG_USERREQ__POST_BARRIER_INIT,
  /* args: Addr barrier, BarrierT type */
  /* To notify the drd tool of a pthread_barrier_destroy call. */
  VG_USERREQ__PRE_BARRIER_DESTROY,
  /* args: Addr barrier, BarrierT type. */
  /* To notify the drd tool of a pthread_barrier_destroy call. */
  VG_USERREQ__POST_BARRIER_DESTROY,
  /* args: Addr barrier, BarrierT type. */
  /* To notify the drd tool of a pthread_barrier_wait call. */
  VG_USERREQ__PRE_BARRIER_WAIT,
  /* args: Addr barrier, BarrierT type. */
  /* To notify the drd tool of a pthread_barrier_wait call. */
  VG_USERREQ__POST_BARRIER_WAIT,
  /* args: Addr barrier, BarrierT type, Word has_waited */

  /* To notify the drd tool of a pthread_rwlock_init call. */
  VG_USERREQ__PRE_RWLOCK_INIT,
  /* args: Addr rwlock */
  /* To notify the drd tool of a pthread_rwlock_destroy call. */
  VG_USERREQ__POST_RWLOCK_DESTROY,
  /* args: Addr rwlock */
  /* To notify the drd tool of a pthread_rwlock_rdlock call. */
  VG_USERREQ__PRE_RWLOCK_RDLOCK,
  /* args: Addr rwlock */
  /* To notify the drd tool of a pthread_rwlock_rdlock call. */
  VG_USERREQ__POST_RWLOCK_RDLOCK,
  /* args: Addr rwlock, Bool took_lock */
  /* To notify the drd tool of a pthread_rwlock_wrlock call. */
  VG_USERREQ__PRE_RWLOCK_WRLOCK,
  /* args: Addr rwlock */
  /* To notify the drd tool of a pthread_rwlock_wrlock call. */
  VG_USERREQ__POST_RWLOCK_WRLOCK,
  /* args: Addr rwlock, Bool took_lock */
  /* To notify the drd tool of a pthread_rwlock_unlock call. */
  VG_USERREQ__PRE_RWLOCK_UNLOCK,
  /* args: Addr rwlock */
  /* To notify the drd tool of a pthread_rwlock_unlock call. */
  VG_USERREQ__POST_RWLOCK_UNLOCK
  /* args: Addr rwlock, Bool unlocked */

};

typedef enum
{
   mutex_type_invalid_mutex    = 0,
   mutex_type_recursive_mutex  = 1,
   mutex_type_errorcheck_mutex = 2,
   mutex_type_default_mutex    = 3,
   mutex_type_spinlock         = 4
} MutexT;

typedef enum
  {
    pthread_barrier = 1,
    gomp_barrier = 2
  } BarrierT;

#endif //  __DRD_CLIENTREQ_H
