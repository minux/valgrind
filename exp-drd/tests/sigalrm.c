#include <assert.h>
#include <errno.h>
#include <pthread.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <asm/unistd.h>
#include "../drd_clientreq.h"


static int s_debug = 0;


static int getktid()
{
  return syscall(__NR_gettid);
}

static int getvgtid()
{
  int res;
  VALGRIND_DO_CLIENT_REQUEST(res, 0, VG_USERREQ__GET_THREAD_SELF, 0, 0, 0,0,0);
  return res;
}

static void SignalHandler(const int iSignal)
{
  if (s_debug)
  {
    char msg[256];
    snprintf(msg, sizeof(msg), "Signal %d was delivered to kernel thread ID %d"
           " / Valgrind thread ID %d\n",
             iSignal, getktid(), getvgtid());
    write(STDOUT_FILENO, msg, strlen(msg));
  }
}

void* thread_func(void* thread_arg)
{
  struct timespec tsRemain, tsDelay;

  if (s_debug)
  {
    printf("thread: kernel thread ID %d  / Valgrind thread ID %d\n",
	   getktid(), getvgtid());
  }

  tsDelay.tv_sec = 10;
  tsDelay.tv_nsec = 0;
  clock_nanosleep(CLOCK_MONOTONIC, 0, &tsDelay, &tsRemain);
  //assert(result < 0 && errno == EINTR);

  return 0;
}

int main(int argc, char** argv)
{
  int vgthreadid;
  pthread_t threadid;
  struct timespec tsDelay;

  // Primitive argument parsing.
  if (argc > 1)
    s_debug = 1;

  vgthreadid = getvgtid();

  if (s_debug)
  {
    printf("main: kernel thread ID %d / Valgrind thread ID %d\n",
	   getktid(), vgthreadid);
  }

  {
    struct sigaction sa;
    memset(&sa, 0, sizeof(sa));
    sa.sa_handler = &SignalHandler;
    sigemptyset(&sa.sa_mask);
    sigaction(SIGALRM, &sa, 0);
  }

  pthread_create(&threadid, 0, thread_func, 0);
  // Wait until the thread is inside clock_nanosleep().
  tsDelay.tv_sec = 0;
  tsDelay.tv_nsec = 20 * 1000 * 1000;
  clock_nanosleep(CLOCK_MONOTONIC, 0, &tsDelay, 0);
  // And send SIGALRM to the thread.
  pthread_kill(threadid, SIGALRM);
  pthread_join(threadid, 0);

  return 0;
}
