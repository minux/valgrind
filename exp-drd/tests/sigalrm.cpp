#include <cassert>
#include <cerrno>
#include <cstdio>
#include <cstdlib>
#include <ctime>
#include <iostream>
#include <pthread.h>
#include <signal.h>
#include <unistd.h>
#include "../drd_clientreq.h"
#include <asm/unistd.h>


#define VALGRIND_START_NEW_SEGMENT    \
{                                                                       \
  int res;                                                              \
  VALGRIND_DO_CLIENT_REQUEST(res, 0, VG_USERREQ__DRD_START_NEW_SEGMENT, \
                             pthread_self(), 0, 0,0,0);                 \
}


static bool s_debug = false;


static int getktid()
{
  return syscall(__NR_gettid);
}

static int getvgtid()
{
  int res;
  VALGRIND_DO_CLIENT_REQUEST(res, 0, VG_USERREQ__GET_THREAD_SELF, 0, 0, 0,0,0);
  return res;
}

static void SignalHandler(const int iSignal)
{
  if (s_debug)
  {
    char msg[256];
    snprintf(msg, sizeof(msg), "Signal %d was delivered to kernel thread ID %d"
           " / Valgrind thread ID %d\n",
             iSignal, getktid(), getvgtid());
    write(STDOUT_FILENO, msg, strlen(msg));
  }
}

void* thread_func(void*)
{
  if (s_debug)
  {
    std::cout << "thread: kernel thread ID " << getktid()
              << " / Valgrind thread ID " << getvgtid() << "\n";
  }

  const timespec tsDelay = { 10, 0 };
  timespec tsRemain;
  clock_nanosleep(CLOCK_MONOTONIC, 0, &tsDelay, &tsRemain);
  //assert(result < 0 && errno == EINTR);

  return 0;
}

int main(int argc, char** )
{
  // Primitive argument parsing.
  if (argc > 1)
    s_debug = true;

  const int vgthreadid = getvgtid();

  if (s_debug)
  {
    std::cout << "main: kernel thread ID " << getktid()
              << " / Valgrind thread ID " << vgthreadid << std::endl;
  }

  {
    struct sigaction sa;
    memset(&sa, 0, sizeof(sa));
    sa.sa_handler = &SignalHandler;
    sigemptyset(&sa.sa_mask);
    sigaction(SIGALRM, &sa, 0);
  }

  pthread_t threadid;
  pthread_create(&threadid, 0, thread_func, 0);
  // Wait until the thread is inside clock_nanosleep().
  const timespec tsDelay = { 0, 20 * 1000 * 1000 };
  clock_nanosleep(CLOCK_MONOTONIC, 0, &tsDelay, 0);
  // And send SIGALRM to the thread.
  pthread_kill(threadid, SIGALRM);
  pthread_join(threadid, 0);

  return 0;
}
