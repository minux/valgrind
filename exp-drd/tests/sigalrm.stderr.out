exp-drd, a data race detector.

Conflicting load by thread 1 at 0x........ size 8
   at 0x........: (within /home/sewardj/VgTRUNK/trunk/exp-drd/tests/sigalrm)
   by 0x........: main (sigalrm.cpp:?)
Allocation context: unknown
Other segment start (thread 2)
   at 0x........: clone (in /...libc...)
   by 0x........: (within libpthread-?.?.so)
   by 0x........: ???
Other segment end (thread 2)
   at 0x........: clock_nanosleep (in /lib64/librt-2.5.so)
   by 0x........: thread_func(void*) (sigalrm.cpp:?)
   by 0x........: vg_thread_wrapper (drd_preloaded.c:?)
   by 0x........: start_thread (in libpthread-?.?.so)
   by 0x........: clone (in /...libc...)

Thread 2:
Conflicting load by thread 2 at 0x........ size 8
   at 0x........: SignalHandler(int) (sigalrm.cpp:?)
Allocation context: stack of thread 2, offset -6320
Other segment start (thread 1)
   at 0x........: clone (in /...libc...)
   by 0x........: do_clone (in libpthread-?.?.so)
   by 0x........: pthread_create@@GLIBC_2.2.5 (in libpthread-?.?.so)
   by 0x........: pthread_create@* (drd_preloaded.c:?)
   by 0x........: main (sigalrm.cpp:?)
Other segment end (thread 1)
   at 0x........: pthread_join (in libpthread-?.?.so)
   by 0x........: pthread_join (drd_preloaded.c:?)
   by 0x........: main (sigalrm.cpp:?)

ERROR SUMMARY: 2 errors from 2 contexts (suppressed: 0 from 0)
